### ISSUE TITLE
Help - Public websocket from existing Backend to the world

### ISSUE THREAD

**Issue by leovinova** [aws/copilot-cli#4131]
Hi,

I have a system, in there: 
- API (Backend service, private subnets, multi AZ)
- Frontend (Public Load Balancer) integrated S3, CloudFront, Route53 
- CMS (Public Load Balancer) integrated S3, CloudFront, Route53

All APIs are routed via `/api/*` paths from my domain 
it works just fine! 

But after added web socket to backend i'm unable to public web socket to internet. Based on [document](https://aws.github.io/copilot-cli/docs/developing/internal-albs/), 
It just supports for `new Backend Service that has not yet been deployed`. 

So I tried to implement an addon to do it.
Basically, i would to: 
- Create `/socket.io` and `/socket.io/*` behaviors  on CloudFront to route web socket request connection and do pub/sub to new target group. Those behaviors SHOULD NOT be cached and SHOULD forward all headers to origin. 
- Create new target group to receive upcoming requests and send it to backend containers 
- Add a Load Balancer rule to forward all requests from CloudFront to new target group

Is that best approach for now? If that please help me to write the addon for the task. 
If not, kindly provide another way to achieve that goal. 
Thanks in advance!

**Comment by huanjani** [comment]
Hello, @leovinova!

Do you need websocket traffic to go from the backend service to the internet only? Or do you also need to receive websocket traffic from the internet to the backend service? 

Thanks!

**Comment by leovinova** [comment]
Hi @huanjani,
I need traffic go out from backend service to the internet only.

**Comment by huanjani** [comment]
Hmmm, then it seems to me that you don't actually need a load balancer for your backend service 🤔.

**Comment by leovinova** [comment]
@huanjani By the way, when you mentioned `websocket traffic to go from the backend service to the internet only`, it means clients from internet connect to my backend websocket right?
if that, we are on the same page 😄 

Thank you for support, I'm looking forward to deployment script, addons... whatever that help me out. 
Much appreciate!

**Comment by huanjani** [comment]
It looks like our docs are out-of-date; thank you for bringing that to our attention! Previously, it wasn't possible to update an existing ECS Service by adding an ALB, but it looks like it is now!
In your backend service manifest, add:
```
http:
  path: '/'
network:
  vpc:
    placement: private
```
then rerun `copilot svc deploy`. That should add an internal ALB to the environment that you're deploying the backend service in.

Let me know how that goes!

**Comment by leovinova** [comment]
my copilot version: 1.21.1
my configuration: 

```sh
http:
  path: "/socket.io"
  healthcheck:
    path: "/hc"
    success_codes: "200"
    healthy_threshold: 3
    unhealthy_threshold: 2
    interval: 15s
    timeout: 3s
  stickiness: true

network:
  vpc:
    placement: private
```

Error when deloy: 
<img width="1054" alt="Screen Shot 2022-11-03 at 13 36 27" src="https://user-images.githubusercontent.com/114042840/199661307-7c0afd04-c13a-4b27-8a53-65a589d809f6.png">

I think internal domain should be `svc.env.app.local` instead of `svc.env.app.internal`, isn't it? @huanjani 

I see `*.local` internal domain in my Route53 hosted zone, but `*.internal` one.

**Comment by leovinova** [comment]
after upgraded to copilot 1.23.0
then added 
```sh
http:
  alias: mydomain.name
```
 got another error

```sh
✘ upload deploy resources for service api: read custom resources for a "Backend Service": read custom resource DynamicDesiredCountFunction at path custom-resources/desired-count-delegation.js: read template custom-resources/desired-count-delegation.js: open templates/custom-resources/desired-count-delegation.js: file does not exist
```
@huanjani

**Comment by huanjani** [comment]
@leovinova, I wonder if you could use a Load Balanced Web Service instead of a Backend Service?

**Comment by leovinova** [comment]
@huanjani yeah, also works for me too. 
As I mentioned, I've implemented it this way manually: 

- Create /socket.io and /socket.io/* behaviors on CloudFront to route web socket request connection and do pub/sub to new target group. Those behaviors SHOULD NOT be cached and SHOULD forward all headers to origin.
- Create new target group to receive upcoming requests and send it to backend containers
- Add a Load Balancer rule to forward all requests from CloudFront to new target group

So how could I write an addon for this? because I need to deploy to various environments as well?

Mainly, when writing addon, I stuck at getting Discovery service IPs into new TargetGroup
Followed is my cloudformation script 

```sh
Resources:
  WebSocketTargetGroup:
    Type: AWS::ElasticLoadBalancingV2::TargetGroup
    Properties:
      HealthCheckEnabled: true
      HealthCheckIntervalSeconds: 15
      HealthCheckPath: /hc
      HealthCheckPort: 8000
      HealthCheckProtocol: HTTP
      HealthCheckTimeoutSeconds: 5
      HealthyThresholdCount: 2
      Matcher: 200
      Port: 80
      Protocol: HTTP
      ProtocolVersion: HTTP2
      TargetGroupAttribute:
        Key: stickiness.enabled
        Value: true
      Targets:
          - Id: !GetAtt Service.Id # service is named of resourced for api service in cloudformation slack
            Port: 8000
```

For `Targets` property, how to fill in Discovery Service IPs there?

**Comment by huanjani** [comment]
Wait--did you `brew install/update copilot`? That's the wrong brew formula; it should be `brew install aws/tap/copilot-cli`!

**Comment by leovinova** [comment]
@huanjani thank you for correct my mistake. 
But after uninstalled and then reinstalled latest version, the error is same as v1.21.1 I shown you above 🥲

**Comment by leovinova** [comment]
@huanjani after upgraded to v1.23.0 and changed alias to `svc.env.app.local` I was able to pass that step, but step after that gets error 

<img width="1145" alt="Screen Shot 2022-11-07 at 17 02 56" src="https://user-images.githubusercontent.com/114042840/200283609-1275163d-341a-495a-8d30-1648015105f2.png">

Bellowed is full my additional config 
```sh
http:
  alias: svc.env.app.local
  path: "/socket.io"
  healthcheck:
    path: "/hc"
    success_codes: "200"
    healthy_threshold: 3
    unhealthy_threshold: 2
    interval: 15s
    timeout: 3s
  stickiness: true

network:
  vpc:
    placement: private
```

**Comment by dannyrandall** [comment]
Hey @leovinova! Based on this requirement you mentioned - "it means clients from internet connect to my backend websocket" - It sounds like this service should be a Load Balanced Web Service. If you use that type, Copilot will set up a public load balancer and a target group to your container for you. We also [recently launched a `cdn` feature](https://aws.github.io/copilot-cli/docs/manifest/environment/#cdn) that creates a CloudFront distribution in front of Load Balanced Web Services for you.

However, since it sounds like you're manually adding this service as an origin on your CloudFront distribution, I think I see how a Backend Service could work.

First, I want to clarify the DNS records: When you add `http:` to a Backend Service: Copilot creates an internal load balancer and a DNS record `svc.app.env.internal` pointing to the _load balancer_. The `.local` address is a service discovery DNS record pointed directly at your ECS tasks. You _should not_ set the `alias` field to either of those values - they'll automatically be added by Copilot, in addition to any aliases you configure. Only use the `alias` field if [you've imported a custom domain](https://aws.github.io/copilot-cli/docs/developing/domain/) that you'd like to use for the service.

One other note is that if you're wanting traffic to go `user -> cloudfront -> internal ALB -> ecs task`, Copilot by default only allows inbound traffic to the internal ALB from the Environment security group. If you want to allow traffic from CloudFront, you'll have to specifically allow inbound traffic to the ALB from [the CloudFront prefix list](https://aws.amazon.com/about-aws/whats-new/2022/02/amazon-cloudfront-managed-prefix-list/).

Does this help clarify things? I'm happy to answer any other questions to help get your service working!

**Comment by leovinova** [comment]
Hey @dannyrandall Thanks for clarification.

Turns out, I've been using imported certificates for my domain. So I have to add `alias` and import private certificates info my environments (dev, stg, prod, etc...) to use Internal Load Balancer

Since I deployed my service used Copilot v1.3.0, there is no option to imported private certs yet. I'll add cert ARN manually  to my manifest file. 

Let's see if I can sort it out.

**Comment by dannyrandall** [comment]
Oh, gotcha! If you imported a cert, then you can use any valid alias for that cert in the `alias` field. For example, if your cert has the names:
```
example.com
*.example.com
```

Then you could set `alias` to `example.com`, `api.example.com`, `internal.example.com`, etc.

**Comment by leovinova** [comment]
@dannyrandall wonderful, 
I was able to create Internal Load Balancer for my Backend Service 

> If you want to allow traffic from CloudFront, you'll have to specifically allow inbound traffic to the ALB from [the CloudFront prefix list](https://aws.amazon.com/about-aws/whats-new/2022/02/amazon-cloudfront-managed-prefix-list/).

But don't know how to get this part done using Copilot. 
It suppose to be divided into 2 smaller parts right? 
- Add Internal Load Balancer CIDRs to the CloudFront prefix list 
- Route request from specified path (i.e: `/socket.io`) to my Internal Load Balancer

**Comment by leovinova** [comment]
Oops! 
I've tried to delete internal load balancer by removed `http` block and deploy. 
Now my stack got `UPDATE_ROLLBACK_FAILED` status. 

Some drifted resources found.
I tried to solve it by clicking on `Stack action` → `Continue update rollback` → `Advance troubleshooting` → skip all resources. Finally, click on `Continue update rollback` button. it FAILED again with message `Only the resources failed during UpdateRollback are allowed to be skipped` 😨 

My Update button also disabled as screenshot
<img width="1486" alt="Screen Shot 2022-11-10 at 18 16 12" src="https://user-images.githubusercontent.com/114042840/201077496-9ad26aa7-f40b-4407-8675-5237cb25dc38.png">

Now I'm in panicking, don't know what to do next @dannyrandall 


Belowed is log for first time failure 

```console
The following resource(s) failed to update: [Service].                                                                                       
  - Update your environment's shared resources                                                                  [update complete]              [105.0s]
    - Request and validate an ACM certificate for your domain                                                   [delete complete]              [34.2s]
    - A load balancer listener to route HTTPS traffic                                                           [update complete]              [5.8s]
    - An internal load balancer listener to route HTTP traffic                                                  [delete complete]              [1.9s]
    - Add an A-record to the hosted zone for the domain alias                                                   [delete complete]              [38.9s]
    - An internal load balancer listener to route HTTPS traffic                                                 [delete complete]              [1.9s]
    - An internal Application Load Balancer to distribute private traffic from within the VPC to your services  [delete complete]              [3.3s]
    - A security group for your internal load balancer allowing HTTP traffic from within the VPC                [delete complete]              [33.7s]
  - An ECS service to run and maintain your tasks in the environment cluster                                    [update failed]                [0.0s]
    Resource handler returned message: "Invalid request provided: UpdateSe                                                                     
    rvice error: The target group with targetGroupArn arn:aws:elasticloadb                                                                     
    alancing:<region>:<account_id>:targetgroup/Targe-36VK3N9YT                                                                     
    Y3C/a3366582ff94927c does not have an associated load balancer. (Servi                                                                     
    ce: AmazonECS; Status Code: 400; Error Code: InvalidParameterException                                                                     
    ; Request ID: da32c025-347c-481c-ada3-a2a712304fd9; Proxy: null)" (Req                                                                     
    uestToken: e0ad298c-e945-314d-71bf-28591987ad9f, HandlerErrorCode: Inv                                                                     
    alidRequest)     
```

**Comment by dannyrandall** [comment]
Hey @leovinova! Is the failure you posted the error from deleting the `http` block? Also, what changes did you make to the load balancer/target group between the time you deployed the service with `http` successfully and when you deleted it?

**Comment by dannyrandall** [comment]
I think in general for your use case, I would recommend using a Load Balanced Web Service and using [the `cdn` feature](https://aws.github.io/copilot-cli/docs/manifest/environment/#cdn) to have Copilot set up CloudFront for you. That should help avoid the situation of needing to update resources outside of CloudFormation that is likely causing the resource drift.

Is there a reason you'd like to use a Backend Service instead of a Load Balanced Web Service, or something that the `cdn` feature is missing for your use case right now?

**Comment by leovinova** [comment]
Hi @dannyrandall 
> Is the failure you posted the error from deleting the http block?

yes it is 

> Also, what changes did you make to the load balancer/target group between the time you deployed the service with http successfully and when you deleted it?

nothing, i just work on copilot 
but, maybe 1 of my team members removed  the Internal Load Balancer listener rule. Because error said that 
I solved the problem already by Update Stack template file manually.
Now my stack hang in `UPDATE_IN_PROGRESS` status forever 🤣  because of waiting ECS service timeout.

plus, ECS container throws this message for stopped container while Copilot keep running infinity.
```console
Stopped reason: Service dev-api-Service-9mXxxiuxxxQU: task remained in deregistered state for too long.
```
> Is there a reason you'd like to use a Backend Service instead of a Load Balanced Web Service, or something that the cdn feature is missing for your use case right now?

my websocket is created inside backend source code (Nodejs 16x). So I think I have to public an internal load balancer as you suggested to internet. 

Should I create new `Load Balanced Web Service` with my backend source code for websocket?

Another question, now my `Backend Service` drifted, can I just simply remove that stack and corresponding sub stacks, then deploy system again?

**Comment by huanjani** [comment]
Hello again, @leovinova!
Yes, we recommend recreating your service, using a `Load Balanced Web Service` as the workload type. The difference between the two types is that LBWSs are internet-facing, and "Backend" services are not. See more here: https://aws.github.io/copilot-cli/docs/concepts/services/#choosing-a-service-type. You may use the same source code. 
We believe you will be able to more easily achieve what you're trying to do by starting with a LBWS. Once you get that running, try `copilot svc delete` for the old Backend Service; if it doesn't work because you're still stuck in `UPDATE_IN_PROGRESS` status, you can try to manually delete those stacks in the CloudFormation console.
Hope this helps!

**Comment by leovinova** [comment]
@huanjani unfortunately, my current frontend(Load Balanced Web Service) is discovering my backend by `.local` domain. 
If i turn my `Backend Service` to `Load Balanced Web Service` then how can i call my api?

**Comment by leovinova** [comment]
and @huanjani, just for your awareness and anyone who read this message at this moment, by my check, if i added Internal Load Balancer into our backend, then DON'T remove it in any case. Because it leads to many troubles later, what i'm facing for now, the root cause is Copilot didn't delete it completely, there are few services left, Target Group for exam. 
In case you would remove Internal Load Balancer, then remove service and re-create it again.

**Comment by Lou1415926** [comment]
> unfortunately, my current frontend(Load Balanced Web Service) is discovering my backend by .local domain.
If i turn my Backend Service to Load Balanced Web Service then how can i call my api?

Hello @leovinova . I am sorry about the churn!

A load-balanced web service is reachable not only over internet via "https://, but also via service discovery endpoint. This means that if you turn your "backend" into a LBWS, it should still be reachable via "backend.env.app.local:80".

> what i'm facing for now, the root cause is Copilot didn't delete it completely, there are few services left, Target Group for exam.

I am happy  to discuss this further and see what was happening there after we get everything working for you! Have you been able to deploy your "backend" as a LBWS?

**Comment by leovinova** [comment]
@Lou1415926 unfortunately, I can't make it too. 
Our database is in secure subnets, it means only private subnets are able to connect. LBWS is in public subnets, so no way i can connect to database. 
Any idea for that?  
Maybe I should go with Amplify + AppSync + Dynamo/Postgres Aurora 🤔 (I had Postgres Aurora already)
or API Gateway web socket API?

**Comment by Lou1415926** [comment]
Happy to help!

> Our database is in secure subnets

Can I get some clarification on what you meant by secure subnets? Do you mean the same thing as private subnets, or are they, for example, specific private subnets with restrictive ACL?

If the database is placed in private subnets, your services in the public subnets should still be able to talk to it - as long as the security group to access the db is configured correctly.

You can refer to the Aurora serverless v2 template that Copilot generates after you run `copilot storage init`; specifically, take a look at the two `"XXXSecurityGroup"` resources and the `"XXXSubnetGroup"` resource. These configuration places the aurora serverless cluster inside of private subnets, and the security group configurations allow the service in public subnet to talk to it.

**Comment by leovinova** [comment]
@Lou1415926 by mentioned `secure subnets`, it's exactly private subnets but just allows only private subnets to connect to it. This design is fixed for security reasons for now. Do you think I need that kind of secure subnets? 

I'm going to take a look into our design from now on. To see if we can remove that barrier.

**Comment by Lou1415926** [comment]
@leovinova By default, the "environment security group" that Copilot places in front of the instance denies traffic from the internet. Therefore, even when you place your service in the public subnets, they are not directly accessible from the internet. For example - suppose you don't have CloudFront enabled -  the traffic goes `Internet -> Application Load Balancer -> LBWS`. 

While I think it is not uncommon for an ECS task in public subnets to talk to a DB in private subnets, if you have specific security compliances, you can also try placing the LBWS in the private subnets by specifying [`placement`](https://aws.github.io/copilot-cli/docs/manifest/lb-web-service/#network-vpc-placement) in the manifest:
```yaml
network:
  vpc:
    placement: 'private'
``` 

The LBWS can still receive internet traffic through the ALB, and it is placed in private subnets.

**Comment by leovinova** [comment]
thanks @Lou1415926 I've chosen to assign security group to my web socket ALBS. And continuously review my security compliant.

**Comment by leovinova** [comment]
@Lou1415926 I found that there is no way to change `network.vpc.security_groups` along with environments change. 
For exam: 
I need to add `sg-aaaaaaaaaaaaa` to `dev` environment, 
so i set 
```console
network:
  vpc:
    placement: 'private'
    security_groups: [sg-aaaaaaaaaaaaa]
```
in my service's manifest file. 

After that, i would deploy service to `staging` with `sg-bbbbbbbbbbbb`. 
How can I tell service to replace `sg-aaaaaaaaaaaaa` by `sg-bbbbbbbbbbbb` programmatically?

**Comment by Lou1415926** [comment]
@leovinova You can use environment override to specify different security groups when the service is deployed to different envs:
```yaml
environments:
  dev:
    network:
      vpc:
        security_groups: sg-aaa
  staging:
    network:
      vpc:
        security_groups: sg-bbb
```

**Comment by leovinova** [comment]
@Lou1415926 well, perfect!

**Comment by leovinova** [comment]
thanks, I'll close this issue since I was able to get it up and running, by creating new ALBS service with code from my api project and attach/create the security group to both RDS and web socket. 
I know this is a workaround but the best approach for now, i think. 

Thanks for quick response with care from the Copilot support team. Good service 👍
